dev/com.ibm.ws.ejbcontainer/src/com/ibm/ws/ejbcontainer/osgi/internal/EJBRuntimeImpl.java

```diff
@@ -242,6 +244,7 @@ private static RuntimeException rethrow(Throwable t) {
 
     private WSEJBHandlerResolver webServicesHandlerResolver;
     private EJBPMICollaboratorFactory ejbPMICollaboratorFactory;
+    private CheckpointPhase checkpointPhase;
```

Comment:
@tjwatson  I made the reference dynamic in case Checkpoint/Restore ever wanted to support application restart, where the app restart would behave normally as without checkpoint.  That is... if the CheckpointPhase could be "unset" after restore, then any subsequent application restarts would restart faster.

This could also be done with a restore hook; but that would require every component that does something similar to have a restore hook.

The UFO doesn't seem to address application restart; while not an intended scenario, would be nice if it clarified the expected behavior.  Is the intended behavior that any new or restarted application after a restore should behave exactly as if a checkpoint is being taken?

Reply:
I don't think I'll want to have the CheckpointPhase service unregistered on the restore side to be able to detect if we have restored.  I think I would prefer a method on the CheckpointPhase (e.g. `boolean restored()`) that returns true once the process has restored.  Two reasons for this.

1) I want the CheckpointPhase to either always there or never there to simplify injection into components (can use static optoinal constructor injection).
2) Allows easy introspection on a dump to see if the checkpoint service is there.

Additionally we can have the CheckpointPhase service have a service property that indicates if it is the restore side.  Then if components would rather do what you are with dynamic optional references then they can use a target filter like:

```java
    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, target="(io.openliberty.checkpoint.restored=false)")
```
This would then cause the reference to be unset on the restore side when we set the service property to true for restored.

Reply:
@tjwatson Thanks for that explanation. Agree there is no reason to make the reference dynamic now; so have changed to use the `@Activate` constructor; final variable.